src/dotty/tools/dotc/transform/ClassTags.scala

```diff
@@ -0,0 +1,57 @@
+package dotty.tools.dotc
+package transform
+
+import core._
+import TreeTransforms._
+import Contexts.Context
+import Flags._
+import SymUtils._
+import Symbols._
+import SymDenotations._
+import Types._
+import Decorators._
+import DenotTransformers._
+import StdNames._
+import NameOps._
+import ast.Trees._
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.core.Constants.Constant
+import util.Positions._
+import Names._
+import collection.mutable
+
+/** This phase replaces calls to DottyPredef.classTag by code that synthesizes appropriate ClassTag
+  */
+class ClassTags extends MiniPhaseTransform with IdentityDenotTransformer { thisTransform =>
+  import ast.tpd._
+
+  object names {
+    val classTag = "classTag".toTermName
+    val typeTag  = "typeTag".toTermName
+  }
+
+  private var classTagCache: Symbol = null
+  private var typeTagCache: Symbol  = null
+  private var scala2ClassTagModule: Symbol = null
+
+
+  override def prepareForUnit(tree: tpd.Tree)(implicit ctx: Context): TreeTransform = {
+    val predefClass = defn.DottyPredefModule.moduleClass.asClass
+    classTagCache = ctx.requiredMethod(predefClass,  names.classTag)
+    typeTagCache = ctx.requiredMethod(predefClass,  names.typeTag)
+    scala2ClassTagModule = ctx.requiredModule("scala.reflect.ClassTag")
+    this
+  }
+
+  override def phaseName: String = "classTags"
+
+  override def transformTypeApply(tree: tpd.TypeApply)(implicit ctx: Context, info: TransformerInfo): tpd.Tree =
+    if (tree.fun.symbol eq classTagCache) {
+      val tp = tree.args.head.tpe
+      val claz = tp.classSymbol
```

Comment:
In Scala, give a value class `Meter`, `classTag[Meter]` and `classOf[Meter]` return `Meter.class`, not the underlying type.

So I think using Scala erasure might not a little bit too simple. You could instead use the java erasure, which, after [my fix](https://github.com/scala/scala/pull/4490) for @smarter's recent bug report, does not unwrap value classes.

Here's a few test cases:

```scala
object Test { 
  type T = String
  type U

  def test = {

    val a /* : Class[T]                  */ = classOf[T]                        // [Ljava/lang/String;
    val b /* : ClassTag[T]               */ = reflect.classTag[T]               // ClassTag(classOf[java.lang.String])

//  val c /* :                           */ = classOf[T with U]                 // error: class type required but Test.T with Test.U found
    val d /* : ClassTag[T with U]        */ = reflect.classTag[T with U]        // ClassTag(classOf[java.lang.String])

    val e /* : Class[Array[T with U]]    */ = classOf[Array[T with U]]          // [Ljava/lang/String;
    val f /* : ClassTag[Array[T with U]] */ = reflect.classTag[Array[T with U]] // ClassTag(arrayClass(classOf[java.lang.String]))

    val g /* : Class[Meter]              */ = classOf[Meter]                    // [LMeter;
    val h /* : ClassTag[Meter]           */ = reflect.classTag[Meter]           // ClassTag(classOf[Meter])
  }
}

class Meter(val i: Int) extends AnyVal
```